### PR TITLE
Add GET:report datapoint for GA4.

### PR DESCRIPTION
**Pull request by techanvil**
## Summary

Addresses issue:

- #6172 

## Relevant technical choices

Rather than adding test cases for the new protected methods `create_analytics_site_data_request` and `parse_reporting_orderby`, which are protected and therefore not part of the public interface and should ideally not be tested directly in a unit test, I have added tests which get the report and check the report parameters have been correctly generated. These are more complete integration-style tests that provides better coverage than isolating these protected methods.

## PR Author Checklist

- [x] My code is tested and passes existing unit tests.
- [x] My code has an appropriate set of unit tests which all pass.
- [x] My code is backward-compatible with WordPress 5.2 and PHP 5.6.
- [x] My code follows the [WordPress](https://make.wordpress.org/core/handbook/best-practices/coding-standards/) coding standards.
- [x] My code has proper inline documentation.
- [x] I have added a QA Brief on the issue linked above.
- [x] I have signed the Contributor License Agreement (see <https://cla.developers.google.com/>).

---------------

_Do not alter or remove anything below. The following sections will be managed by moderators only._

## Code Reviewer Checklist

- [ ] Run the code.
- [ ] Ensure the acceptance criteria are satisfied.
- [ ] Reassess the implementation with the IB.
- [ ] Ensure no unrelated changes are included.
- [ ] Ensure CI checks pass.
- [ ] Check Storybook where applicable.
- [ ] Ensure there is a QA Brief.

## Merge Reviewer Checklist

- [ ] Ensure the PR has the correct target branch.
- [ ] Double-check that the PR is okay to be merged.
- [ ] Ensure the corresponding issue has a ZenHub release assigned.
- [ ] Add a changelog message to the issue.

### CODE REVIEW

File: tests/phpunit/integration/Modules/Analytics_4Test.php [hunk 1]
```diff
@@ -206,13 +209,373 @@ public function test_get_datapoints() {
...
 				'property',
+				'report',
 				'webdatastreams',
 				'webdatastreams-batch',
 			),
 			$this->analytics->get_datapoints()
 		);
 	}
 
+	public function test_get_report() {
```
Comment (techanvil) [inline tests/phpunit/integration/Modules/Analytics_4Test.php:220]: As described in **Relevant technical choices**, I've opted to write these higher-level tests which provide better coverage rather than individually testing the protected methods as described in the IB.

File: includes/Modules/Analytics_4.php [hunk 2]
```diff
@@ -826,4 +861,468 @@ public function check_service_entity_access() {
...
+	/**
+	 * Validates the report metrics.
+	 *
+	 * @since n.e.x.t
+	 *
+	 * @param Google_Service_AnalyticsData_Metric[] $metrics The metrics to validate.
+	 * @throws Invalid_Report_Metrics_Exception Thrown if the metrics are invalid.
+	 */
+	protected function validate_report_metrics( $metrics ) {
+		if ( false === $this->is_using_shared_credentials ) {
```
Comment (aaemnnosttv) [inline includes/Modules/Analytics_4.php:1226]: Note: the changes in #5780 are relevant here. We should keep this as-is for now of course because that one is still in definition but I'll make a note on that one that the proposed changes there would apply here as well.

File: includes/Modules/Analytics_4.php [hunk 3]
```diff
@@ -588,6 +622,7 @@ protected function setup_services( Google_Site_Kit_Client $client ) {
 		return array(
 			'analyticsadmin' => new Google_Service_GoogleAnalyticsAdmin( $client ),
 			'tagmanager'     => new Google_Service_TagManager( $client ),
+			'analyticsdata'  => new Google_Service_AnalyticsData( $client ),
```
Comment (aaemnnosttv) [inline includes/Modules/Analytics_4.php:625]: 🙂 
```suggestion
			'analyticsdata'  => new Google_Service_AnalyticsData( $client ),
			'tagmanager'     => new Google_Service_TagManager( $client ),
```

File: includes/Modules/Analytics_4.php [hunk 4]
```diff
@@ -826,4 +861,468 @@ public function check_service_entity_access() {
...
+	 * @param Data_Request $data Data request object.
+	 * @return RequestInterface|WP_Error Request object on success, or WP_Error on failure.
+	 */
+	protected function create_report_request( Data_Request $data ) {
+		$request_args = array();
+
+		if ( empty( $data['propertyID'] ) ) {
+			/* translators: %s: Missing parameter name */
+			return new WP_Error( 'missing_required_param', sprintf( __( 'Request parameter is empty: %s.', 'google-site-kit' ), 'propertyID' ), array( 'status' => 400 ) );
+		}
```
[the remark this answers is not in the document]
Comment (techanvil) [reply includes/Modules/Analytics_4.php]: Thanks for pointing this out, I've changed to always use the connected property via settings in the handler.

File: tests/phpunit/integration/Modules/Analytics_4Test.php [hunk 5]
```diff
@@ -206,13 +212,672 @@ public function test_get_datapoints() {
...
+		$user_id        = $this->factory()->user->create( array( 'role' => 'administrator' ) );
+		$context        = new Context( GOOGLESITEKIT_PLUGIN_MAIN_FILE );
+		$options        = new Options( $context );
+		$user_options   = new User_Options( $context, $user_id );
+		$authentication = new Authentication( $context, $options, $user_options );
+		$analytics      = new Analytics_4( $context, $options, $user_options, $authentication );
+
+		$authentication->get_oauth_client()->get_client()->setHttpClient(
+			new FakeHttpClient() // Returns 200 by default.
+		);
```
[the remark this answers is not in the document]
Comment (techanvil) [reply tests/phpunit/integration/Modules/Analytics_4Test.php:747]: Thanks, good spot - this was left in by mistake. Have removed.

File: tests/phpunit/integration/Modules/Analytics_4Test.php [hunk 6]
```diff
@@ -206,13 +212,672 @@ public function test_get_datapoints() {
...
+	 * @param array $report_params The report parameters.
+	 * @param array $request_handler_calls Output variable for tracking the request handler calls. Passed by reference.
+	 * @return RunReportResponse[] The report response.
+	 */
+	protected function get_report( array $report_params, array &$request_handler_calls ) {
+		$property_id = isset( $report_params['propertyID'] ) ? $report_params['propertyID'] : null;
+
+		$analytics = $this->setup_report( $property_id, $request_handler_calls );
+
+		return $analytics->get_data( 'report', $report_params );
```
[the remark this answers is not in the document]
Comment (techanvil) [reply tests/phpunit/integration/Modules/Analytics_4Test.php]: Sure, happy to refactor a bit and I've changed `$request_handler_calls` to be a property as suggested.

File: tests/phpunit/integration/Modules/Analytics_4Test.php [hunk 7]
```diff
@@ -206,13 +212,672 @@ public function test_get_datapoints() {
...
+						);
+
+					default:
+						return new Response( 200 );
+				}
+			}
+		);
+
+		$analytics->get_client()->setHttpClient( $http_client );
+		$analytics->register();
```
[the remark this answers is not in the document]
Comment (techanvil) [reply tests/phpunit/integration/Modules/Analytics_4Test.php:810]: I've extracted the HTTP client creation to a separate helper method :+1:

File: tests/phpunit/integration/Modules/Analytics_4Test.php [hunk 8]
```diff
@@ -206,13 +212,672 @@ public function test_get_datapoints() {
...
+
+			add_option(
+				Module_Sharing_Settings::OPTION,
+				array(
+					'analytics-4' => array(
+						'sharedRoles' => array( 'administrator' ),
+						'management'  => 'all_admins',
+					),
+				)
+			);
```
[the remark this answers is not in the document]
Comment (techanvil) [reply tests/phpunit/integration/Modules/Analytics_4Test.php:853]: I've extracted the shared credentials setup to a separate helper method to call directly from the test cases :+1:

File: tests/phpunit/integration/Modules/Analytics_4Test.php [hunk 9]
```diff
@@ -206,13 +212,672 @@ public function test_get_datapoints() {
...
+			}
+		);
+
+		$analytics->get_client()->setHttpClient( $http_client );
+		$analytics->register();
+
+		if ( $enable_validation ) {
+			// Metrics and dimensions are only validated when using shared credentials; this block of code sets up the shared credentials scenario.
+
+			$this->enable_feature( 'dashboardSharing' );
```
[the remark this answers is not in the document]
Comment (techanvil) [reply tests/phpunit/integration/Modules/Analytics_4Test.php:815]: I've reorganised the relevant tests to enable this at the start of them.

File: tests/phpunit/integration/Modules/Analytics_4Test.php [hunk 10]
```diff
@@ -206,13 +212,672 @@ public function test_get_datapoints() {
...
+		return $analytics->get_data( 'report', $report_params );
+	}
+
+	/**
+	 * Sets up a mock Analytics 4 instance in preparation for retrieving a report.
+	 *
+	 * This is a helper method to avoid duplicating the same code in multiple tests.
+	 * It also allows us to mock the HTTP client to verify the request parameters.
+	 *
+	 * @since n.e.x.t
```
[the remark this answers is not in the document]
Comment (techanvil) [reply tests/phpunit/integration/Modules/Analytics_4Test.php:731]: I'd agree they aren't really needed here, have removed :+1:

File: tests/phpunit/integration/Modules/Analytics_4Test.php [hunk 11]
```diff
@@ -206,13 +212,672 @@ public function test_get_datapoints() {
...
+						return new Response( 200 );
+				}
+			}
+		);
+
+		$analytics->get_client()->setHttpClient( $http_client );
+		$analytics->register();
+
+		if ( $enable_validation ) {
+			// Metrics and dimensions are only validated when using shared credentials; this block of code sets up the shared credentials scenario.
```
[the remark this answers is not in the document]
Comment (techanvil) [reply tests/phpunit/integration/Modules/Analytics_4Test.php:813]: Have extracted to a helper method as mentioned. Also, yeah I do prefer to avoid boolean parameters, should have done so here really. It's gone now anyway!

File: tests/phpunit/integration/Modules/Analytics_4Test.php [hunk 12]
```diff
@@ -206,13 +212,672 @@ public function test_get_datapoints() {
...
+					'totalUsers',
+				);
+			}
+		);
+
+		$property_id = '123456789';
+
+		$request_handler_calls = array();
+
+		$analytics = $this->setup_report( $property_id, $request_handler_calls, true );
```
[the remark this answers is not in the document]
Comment (techanvil) [reply tests/phpunit/integration/Modules/Analytics_4Test.php]: Following our discussion I've taken an approach using a `@dataProvider` to provide an access token (or not) for each test, not entirely convinced about it, please take a look and see what you think...

File: tests/phpunit/integration/Modules/Analytics_4Test.php [hunk 13]
```diff
@@ -206,13 +212,672 @@ public function test_get_datapoints() {
...
+					'name' => 'sessionDefaultChannelGrouping',
+				),
+				array(
+					'name' => 'pageTitle',
+				),
+			),
+			$request_params['dimensions']
+		);
+
+		$this->assertEquals(
```
[the remark this answers is not in the document]
Comment (techanvil) [reply tests/phpunit/integration/Modules/Analytics_4Test.php:352]: As discussed it turns out the `assertEquals` is actually key-order agnostic so it looks like we can leave these as-is. I think PHPUnit is using this [`Comparator`](https://github.com/sebastianbergmann/comparator/blob/ef3a88a13a48f3db94e1c7af4ba090819d3aa6cd/src/ArrayComparator.php#L19-L26) class under the hood...

File: tests/phpunit/integration/Modules/Analytics_4Test.php [hunk 14]
```diff
@@ -206,13 +212,672 @@ public function test_get_datapoints() {
...
+	 * @param boolean [enable_validation] Whether to enable validation of the metrics and dimensions. Default false.
+	 * @return Analytics_4 The Analytics 4 instance.
+	 */
+	protected function setup_report( $property_id, array &$request_handler_calls, $enable_validation = false ) {
+		$user_id        = $this->factory()->user->create( array( 'role' => 'administrator' ) );
+		$context        = new Context( GOOGLESITEKIT_PLUGIN_MAIN_FILE );
+		$options        = new Options( $context );
+		$user_options   = new User_Options( $context, $user_id );
+		$authentication = new Authentication( $context, $options, $user_options );
+		$analytics      = new Analytics_4( $context, $options, $user_options, $authentication );
```
[the remark this answers is not in the document]
Comment (techanvil) [reply tests/phpunit/integration/Modules/Analytics_4Test.php:743]: Sure, have moved to the `set_up` method :+1:

File: includes/Modules/Analytics_4.php (hunk 3 above)
Comment (techanvil) [reply includes/Modules/Analytics_4.php:623]: Thanks for the spot! Updated.

**Review by aaemnnosttv — COMMENTED** [review]
Thanks @techanvil – there were a few more things I noticed to address here but this is looking great. Almost there I think 👍

File: includes/Modules/Analytics_4.php [hunk 15]
```diff
@@ -37,18 +37,32 @@
...
 use Google\Site_Kit\Core\Util\URL;
+use Google\Site_Kit\Core\Validation\Exception\Invalid_Report_Dimensions_Exception;
+use Google\Site_Kit\Core\Validation\Exception\Invalid_Report_Metrics_Exception;
 use Google\Site_Kit\Modules\Analytics\Settings as Analytics_Settings;
 use Google\Site_Kit\Modules\Analytics_4\Settings;
 use Google\Site_Kit\Modules\Analytics_4\Tag_Guard;
 use Google\Site_Kit\Modules\Analytics_4\Web_Tag;
 use Google\Site_Kit_Dependencies\Google\Model as Google_Model;
+use Google\Site_Kit_Dependencies\Google\Service\AnalyticsData as Google_Service_AnalyticsData;
+use Google\Site_Kit_Dependencies\Google\Service\AnalyticsData\BatchRunReportsRequest as Google_Service_AnalyticsData_BatchRunReportsRequest;
```
[the remark this answers is not in the document]
Comment (techanvil) [reply includes/Modules/Analytics_4.php]: Thanks for the spot, have removed!

File: includes/Modules/Analytics_4.php [hunk 16]
```diff
@@ -826,4 +859,460 @@ public function check_service_entity_access() {
...
+	protected function create_report_request( Data_Request $data ) {
+		$request_args = array();
+
+		if ( empty( $data['metrics'] ) ) {
+			/* translators: %s: Missing parameter name */
+			return new WP_Error( 'missing_required_param', sprintf( __( 'Request parameter is empty: %s.', 'google-site-kit' ), 'metrics' ), array( 'status' => 400 ) );
+		}
+
+		$option      = $this->get_settings()->get();
+		$property_id = self::normalize_property_id( $option['propertyID'] );
```
[the remark this answers is not in the document]
Comment (techanvil) [reply includes/Modules/Analytics_4.php]: Sounds good, I've updated to return an error when `propertyID` is empty and moved the normalization to the point of use.

I added a test for the empty `propertyID` scenario. I've also refactored the validation tests slightly as I realised they don't actually need the fake HTTP client as they don't get as far as making a request. As a result I noticed another test scenario to cover and added a test for insufficient permissions.

File: includes/Modules/Analytics_4.php [hunk 17]
```diff
@@ -826,4 +859,460 @@ public function check_service_entity_access() {
...
+			$date_range    = $data['dateRange'] ?: 'last-28-days';
+			$date_ranges[] = $this->parse_date_range( $date_range, $data['compareDateRanges'] ? 2 : 1 );
+
+			// When using multiple date ranges, it changes the structure of the response:
+			// Aggregate properties (minimum, maximum, totals) will have an entry per date range.
+			// The rows property will have additional row entries for each date range.
+			if ( ! empty( $data['multiDateRange'] ) ) {
+				$date_ranges[] = $this->parse_date_range( $date_range, 1, 1, true );
+			}
+		}
```
[the remark this answers is not in the document]
Comment (techanvil) [reply includes/Modules/Analytics_4.php]: Sounds good, I've removed `dateRange`, `compareDateRanges`, and `multiDateRange`.

I have, though, kept the default case where no date params are provided at all, in which case it defaults to the last 28 days. I've also added a test case as this scenario was missed in the coverage. If you'd prefer to remove this default entirely, please let me know.

Have also created a followup issue to remove these legacy params from the codebase: https://github.com/google/site-kit-wp/issues/6395

File: includes/Modules/Analytics_4.php [hunk 18]
```diff
@@ -826,4 +859,460 @@ public function check_service_entity_access() {
...
+
+		$dimension_filter_expressions = array();
+
+		$hostnames = array_values(
+			array_unique(
+				array_map(
+					function ( $site_url ) {
+						return URL::parse( $site_url, PHP_URL_HOST );
+					},
+					$this->permute_site_url( $this->context->get_reference_site_url() )
```
[the remark this answers is not in the document]
Comment (techanvil) [reply includes/Modules/Analytics_4.php]: Sounds good, have updated in both.

File: tests/phpunit/integration/Modules/Analytics_4Test.php [hunk 19]
```diff
@@ -206,13 +237,848 @@ public function test_get_datapoints() {
...
+
+		$this->analytics->get_data(
+			'report',
+			array(
+				// Note, metrics is a required parameter.
+				'metrics'        => array(
+					array( 'name' => 'sessions' ),
+				),
+				'dateRange'      => 'last-14-days',
+				'multiDateRange' => true,
```
[the remark this answers is not in the document]
Comment (techanvil) [reply tests/phpunit/integration/Modules/Analytics_4Test.php]: Cheers, have removed :+1:

File: tests/phpunit/integration/Modules/Analytics_4Test.php [hunk 20]
```diff
@@ -206,13 +237,848 @@ public function test_get_datapoints() {
...
+		$this->assertWPErrorWithMessage( 'Request parameter is empty: metrics.', $data );
+	}
+
+	public function test_report__metric_validation() {
+		$this->enable_feature( 'dashboardSharing' );
+
+		$this->context        = new Context( GOOGLESITEKIT_PLUGIN_MAIN_FILE );
+		$options              = new Options( $this->context );
+		$this->user           = $this->factory()->user->create_and_get( array( 'role' => 'administrator' ) );
+		$user_options         = new User_Options( $this->context, $this->user->ID );
```
[the remark this answers is not in the document]
Comment (techanvil) [reply tests/phpunit/integration/Modules/Analytics_4Test.php]: Hmm, I had recreated them all in response to your comment about enabling the feature at the start of the test. I think it's valid to keep these because the fact that these instantiations shouldn't be subject to feature flag changes implies knowledge of their implementation that we arguably should not assume from a testing perspective.

To be honest I think we're doing a lot of this sort of instantiation in test `set_up` methods and then enabling features in the tests elsewhere in the test codebase, so it would not be inconsistent to do so, but also think it's a reasonable call to leave this as-is for the reason outlined above. I've left them as they are.

File: tests/phpunit/integration/Modules/Analytics_4Test.php [hunk 21]
```diff
@@ -206,13 +237,848 @@ public function test_get_datapoints() {
...
+
+	/**
+	 * Returns a date string for the given number of days ago.
+	 *
+	 * @param int $days_ago The number of days ago.
+	 * @return string The date string, formatted as YYYY-MM-DD.
+	 */
+	protected function days_ago_date_string( $days_ago ) {
+		return gmdate( 'Y-m-d', strtotime( $days_ago . ' days ago' ) );
+	}
```
[the remark this answers is not in the document]
Comment (techanvil) [reply tests/phpunit/integration/Modules/Analytics_4Test.php:868]: I've ended up leaving this in, due to the new default date range test case mentioned above.

I started looking for solutions for mocking the date in PHPUnit but didn't want to go too far down a rabbit hole.

Given the low probability of hitting the edge case you mentioned on a given test run, I'd suggest leaving it in for now and following up with a separate issue to provide a date mocking solution and rewrite the test, as this would also provide useful functionality for other tests. What do you think?

File: tests/phpunit/integration/Modules/Analytics_4Test.php [hunk 22]
```diff
@@ -206,13 +237,848 @@ public function test_get_datapoints() {
...
+	/**
+	 * Metrics and dimensions are only validated when using shared credentials. This helper method sets up the shared credentials scenario.
+	 */
+	protected function enable_shared_credentials() {
+		// Ensure the user is authenticated.
+		$this->authentication->get_oauth_client()->set_token(
+			array(
+				'access_token' => 'valid-auth-token',
+			)
+		);
```
[the remark this answers is not in the document]
Comment (techanvil) [reply tests/phpunit/integration/Modules/Analytics_4Test.php]: Good spot! Have updated :+1:

File: tests/phpunit/integration/Modules/Analytics_4Test.php [hunk 23]
```diff
@@ -206,13 +237,848 @@ public function test_get_datapoints() {
...
+				'propertyID'      => '123',
+				'webDataStreamID' => '456',
+				'measurementID'   => 'G-789',
+				'ownerID'         => $this->user->ID,
+			)
+		);
+
+		// Setup a user with shared access to the Analytics 4 module.
+		$admin = $this->factory()->user->create_and_get( array( 'role' => 'administrator' ) );
+		wp_set_current_user( $admin->ID );
```
[the remark this answers is not in the document]
Comment (techanvil) [reply tests/phpunit/integration/Modules/Analytics_4Test.php]: I've made that change, although it required a couple of steps in addition to setting the access token (see d41ea9503776832243a21da8c18c66198c46ef4b).

File: tests/phpunit/integration/Modules/Analytics_4Test.php [hunk 24]
```diff
@@ -206,13 +237,848 @@ public function test_get_datapoints() {
...
+
+		add_option(
+			Module_Sharing_Settings::OPTION,
+			array(
+				'analytics-4' => array(
+					'sharedRoles' => array( 'administrator' ),
```
[the remark this answers is not in the document]
Comment (techanvil) [reply tests/phpunit/integration/Modules/Analytics_4Test.php]: Makes sense, have done :+1:

File: tests/phpunit/integration/Modules/Analytics_4Test.php [hunk 25]
```diff
@@ -206,13 +237,848 @@ public function test_get_datapoints() {
...
+					'management'  => 'all_admins',
```
[the remark this answers is not in the document]
Comment (techanvil) [reply tests/phpunit/integration/Modules/Analytics_4Test.php]: Sounds fair, have done :+1:

File: tests/phpunit/integration/Modules/Analytics_4Test.php [hunk 26]
```diff
@@ -206,13 +237,848 @@ public function test_get_datapoints() {
...
+			'authenticated user'   => array( 'valid-token' ),
+		);
+	}
+
+	/**
+	 * Sets up user authentication if an access token is provided.
+	 *
+	 * @param string $access_token The access token to use.
+	 */
+	protected function setup_user_authentication( $access_token ) {
```
[the remark this answers is not in the document]
Comment (techanvil) [reply tests/phpunit/integration/Modules/Analytics_4Test.php]: Sounds good, I have extracted setting the user token to a new trait, `UserAuthenticationTrait`.

File: tests/phpunit/integration/Modules/Analytics_4Test.php (hunk 20 above)
[the remark this answers is not in the document]
Comment (aaemnnosttv) [reply tests/phpunit/integration/Modules/Analytics_4Test.php]: Sounds good to me 👍

File: includes/Modules/Analytics_4.php [hunk 27]
```diff
@@ -826,4 +858,447 @@ public function check_service_entity_access() {
...
+	 * @param Data_Request $data Data request object.
+	 * @return RequestInterface|WP_Error Request object on success, or WP_Error on failure.
+	 */
+	protected function create_report_request( Data_Request $data ) {
+		$request_args = array();
+
+		$option = $this->get_settings()->get();
+
+		if ( empty( $option['propertyID'] ) ) {
+			return new WP_Error( 'missing_required_setting', __( 'No connected Google Analytics 4 property ID.', 'google-site-kit' ), array( 'status' => 400 ) );
```
Comment (aaemnnosttv) [inline includes/Modules/Analytics_4.php:875]: A bit of a nitpick here but a 400 response would indicate that the request was malformed in some way that it should not be retried. Since this condition is based on internal state, I think a 500 error would be more appropriate, and this probably makes sense to go after the 4xx errors (i.e. once we determine the request was made correctly).

File: tests/phpunit/integration/Modules/Analytics_4Test.php (hunk 21 above)
[the remark this answers is not in the document]
Comment (aaemnnosttv) [reply tests/phpunit/integration/Modules/Analytics_4Test.php:868]: Sure, that sounds good to me. It's very unlikely that the test would fail for this reason, but I agree we should preserve the default date range 👍

File: tests/phpunit/integration/Modules/Analytics_4Test.php [hunk 28]
```diff
@@ -206,13 +247,787 @@ public function test_get_datapoints() {
...
+	 * @param string $property_id The GA4 property ID to use.
+	 * @return FakeHttpClient The fake HTTP client.
+	 */
+	protected function create_fake_http_client( $property_id ) {
+		$this->request_handler_calls = array();
+
+		// Grant scopes so request doesn't fail.
+		$this->authentication->get_oauth_client()->set_granted_scopes(
+			$this->analytics->get_scopes()
+		);
```
Comment (aaemnnosttv) [inline tests/phpunit/integration/Modules/Analytics_4Test.php:910]: Let's pull this part out into the respective tests (or a separate helper) so that this method is specific to the client.

**Review by aaemnnosttv — COMMENTED** [review]
Thanks for your patience on this @techanvil – this is very close, just a few final details to touch up.

File: tests/phpunit/integration/Modules/Analytics_4Test.php (hunk 21 above)
[the remark this answers is not in the document]
Comment (techanvil) [reply tests/phpunit/integration/Modules/Analytics_4Test.php:868]: Ok, cool. I've created the issue #6407 to follow up on mocking the date in PHPUnit tests.

File: includes/Modules/Analytics_4.php (hunk 27 above)
Comment (techanvil) [reply includes/Modules/Analytics_4.php]: Good spot! I've done this, and also updated the metric/dimension validation errors to return a 400 rather than the current default of 500. Also updated the tests to check for the status code.

File: tests/phpunit/integration/Modules/Analytics_4Test.php (hunk 28 above)
Comment (techanvil) [reply tests/phpunit/integration/Modules/Analytics_4Test.php]: No probs, that's done :+1:

File: includes/Modules/Analytics_4.php [hunk 29]
```diff
@@ -926,7 +922,8 @@ function ( $dimension_def ) {
 				} catch ( Invalid_Report_Dimensions_Exception $exception ) {
 					return new WP_Error(
 						'invalid_analytics_4_report_dimensions',
-						$exception->getMessage()
+						$exception->getMessage(),
+						array( 'status' => 400 )
```
Comment (aaemnnosttv) [inline includes/Modules/Analytics_4.php:926]: I can see why you used the status here as well but let's not do that because it would imply we should make the same changes in the other modules that also implement this check. Aside from that, I think the default status is okay here because the client is not making an invalid request – the validation is applied due to internal state/conditions related to dashboard sharing.

**Review by aaemnnosttv — COMMENTED** [review]
Thanks @techanvil ! Just one thing related to one of the extra changes you made. I'll update this myself to avoid another round here as it should be a trivial change to make, but great job here 👏

File: includes/Modules/Analytics_4.php (hunk 27 above)
Comment (aaemnnosttv) [reply includes/Modules/Analytics_4.php]: Thanks! As mentioned in https://github.com/google/site-kit-wp/pull/6379#discussion_r1070000992 I don't think we should change the status of the errors for the DS-related validation because that is also dependent on internal state rather than an improperly made request. I've gone ahead and reverted this part.

File: includes/Modules/Analytics_4.php [hunk 30]
```diff
@@ -831,4 +863,455 @@ public function check_service_entity_access() {
...
+				'dimension_filters' => array(),
+				'start_date'        => '',
+				'end_date'          => '',
+				'page'              => '',
+				'row_limit'         => '',
+			)
+		);
+
+		$request = new Google_Service_AnalyticsData_RunReportRequest();
+		$request->setProperty( self::normalize_property_id( $property_id ) );
```
Comment (aaemnnosttv) [inline includes/Modules/Analytics_4.php:1148]: This seems to be unnecessary but we can look at it in a separate issue as I don't think it's hurting anything here. The property ID is provided as a path parameter in the API request, but the object we're building here is essentially the body of that request so it's being duplicated. It's also not mentioned in the [schema for the request body](https://developers.google.com/analytics/devguides/reporting/data/v1/rest/v1beta/properties/runReport), but is specifiable in the API explorer 😄 🤷‍♂️ 

Perhaps this is a leftover detail from an earlier version of the API or maybe there for consistency with the batch method (although I think a batch is still limited to the same property). See https://github.com/googleapis/google-api-php-client-services/blob/ff7e67045c666d292e07e0898ce63be458592745/src/AnalyticsData/Resource/Properties.php#L199-L205

**Review by aaemnnosttv — APPROVED** [review]
Great! Thanks @techanvil 💪

File: includes/Modules/Analytics_4.php (hunk 29 above)
Comment (techanvil) [reply includes/Modules/Analytics_4.php]: No worries and thanks for changing it back :+1:

File: includes/Modules/Analytics_4.php (hunk 30 above)
Comment (techanvil) [reply includes/Modules/Analytics_4.php:1148]: Yes, possibly unnecessary, worth taking another look at :+1:
